src/sdk/data/helpers/picks-helpers.ts

```diff
@@ -0,0 +1,377 @@
+      }
+    }
+
+    bottom = top
+    if (traverse && header.parent && md > fromMsl) {
+      const parent = headers.get(header.parent)
+      if (parent) await addPicks(parent)
+    }
+  }
+
+  await addPicks(wellbore)
+
+  return { matched, unmatched, wellbore }
+}
+
+/**
+ * Create groups of entry and exit picks for each stratigraphy unit level:
+ * 
+ * - Sort the matched picks by ascending level and ascending depth
+ * - Traverse the matched picks, keeping a reference of the current level and unit to find the correct entry and exit pick
+ * - Remove picks from list as formations are identified
+ */
+export function createFormationIntervals(
+  unitPicks: UnitPick[],
+  maxDepth: number = Infinity,
+): FormationInterval[] {
+  const intervals: FormationInterval[] = []
+
+  if (!Array.isArray(unitPicks) || unitPicks.length < 2) return intervals
+
+  let list = unitPicks.map((up) => ({
+    ...up,
+    deleted: false,
+  })) // make a copy
+
+  list.sort(
+    (a, b) =>
+      a.unit.level - b.unit.level ||
+      a.pick.mdMsl - b.pick.mdMsl ||
+      a.unit.topAge - b.unit.topAge
+  )
+
+  const findNext = () => {
+    let interval = null
+    const entry = list.shift()
+    if (entry) {
+      for (let i = 0; i < list.length; i++) {
+        const candidate = list[i]
+
+        // if we react a new strat unit level, it means we won't find a matching exit pick due
+        // to how we ordered the unit picks
+        if (entry.unit.level < candidate.unit.level) {
+          break
+        }
+
+        if (candidate.unit.name === entry.unit.name) {
+          if (candidate.pick.pickIdentifier !== entry.pick.pickIdentifier) {
+            list[i].deleted = true // mark item as deleted
+            interval = {
+              entry: entry.pick,
+              exit: candidate.pick,
+              unit: entry.unit,
+            }
+
+            break
+          } else {
+            list[i].deleted = true // remove candidate from list as it is a duplicate of the parent wellbore entry pick
+          }
+        }
+      }
+
+      list = list.filter((d) => !d.deleted)
+    }
+
+    return interval
+  }
+
+  while (list.length) {
+    const interval = findNext()
+
+    // Only add interval if it has a range
+    if (interval && interval.entry.mdMsl !== interval.exit.mdMsl) {
+      /*
+      // if the interval is of the same unit as the previous, and there's no gap between them,
+      // modify previous interval instead of adding a new
+      const prevInterval = intervals[intervals.length - 1]
+      if (prevInterval && prevInterval.unit.identifier === interval.unit.identifier && prevInterval.exit.md === interval.entry.md) {
+        prevInterval.exit = { ...prevInterval.exit } // avoid mutating original data
+        prevInterval.exit.md = interval.exit.md
+        prevInterval.exit.mdMsl = interval.exit.mdMsl
+        prevInterval.exit.tvdMsl = interval.exit.tvdMsl
+      } else {
+        intervals.push(interval)  
+      }
+      */
```

Comment:
Are you planning on keeping the commented out code?

Reply:
It will be cleaned up later. We may decide to merge these intervals at some point.